crates/core/src/network/connection_manager/mod.rs

```diff
@@ -505,8 +558,13 @@ where
         MESSAGE_RECEIVED.increment(1);
         let processing_started = Instant::now();
         // header is optional on non-hello messages.
-        if let Some(_header) = msg.header {
-            // todo: if header contains newer config or metadata versions, notify metadata().
+        if let Some(header) = msg.header {
+            metadata_version_reporter.report(
+                header.my_nodes_config_version.map(Into::into),
+                header.my_partition_table_version.map(Into::into),
+                header.my_schema_version.map(Into::into),
+                header.my_logs_version.map(Into::into),
+            );
```

Comment:
I wonder if the following is a worthwhile micro-optimization: Keeping local variables in the loop with the max_observed versions and only report if they are different (no atomics, cache coherence, etc.)

Reply:
I think this is a good improvement. Will implement it.